### PR TITLE
Fix #27

### LINKED ISSUE #27: Installing on Windows?

### PR DESCRIPTION
**Pull request by Justaus3r**
This Pull request should hopefully close #27 .Following changes have been made:
1:Add windows support.
2:Change the default anime provider to ```animerush``` as it gives **far** better search results.
3:Add documentation for installing adl on windows  and improve docs a bit  .

Tested on Windows 7,10,Ubuntu20.4 and Pop!_Os.

### CODE REVIEW

**Comment by RaitaroH** [comment]
Thank you very much for the PR.

I just added some changes to how adl matches local files using perl. Please also test against that on your machine. I will check to compatability of this PR in due time.

Regarding 2, I would preffer to have the config file for anime dl itself take care of that. Doing this way would actually overwrite the config choice. For instance I have `vidstream` as default. I do understand what you are trying to say, the windows users might not change that provider from `twist.moe` and get bad results. But, is also true that providers might stop working at any time. I will think about how to deal with that.

Also, also that imgbot made the svg funny. No problem, I will just have to work around it.

**Comment by Justaus3r** [comment]
>> Thank you very much for the PR.

No problem,it's a great bash script so i wanted it to be usable on windows too.

>> I just added some changes to how adl matches local files using perl. Please also test against that on your machine. I will check to compatability of this PR in due time.

Tested(for now on linux) and its working perfectly fine.

>>Regarding 2, I would preffer to have the config file for anime dl itself take care of that. Doing this way would actually overwrite the config choice. For instance I have vidstream as default. I do understand what you are trying to say, the windows users might not change that provider from twist.moe and get bad results. But, is also true that providers might stop working at any time. I will think about how to deal with that.

So maybe we could check if the config exists then use that ,otherwise use ```animerush```.i didn't only tested on my machine but also on with my friends anilist and ```anilist``` seems to be giving excellent results on both linux and windows.the default provider gives me **teribble**(and i mean really terrible) results.so i would **really** encourage you to use ```animerush``` as default provider.i dont think users really bother with creating custom configs so there's that.also well any provider could get down at anytime but we do have an option in adl for switching providers if the current provider is not working so i don't think its much of a problem.**BUT** at last its your decision as its your repository 🙂.

>>Also, also that imgbot made the svg funny. No problem, I will just have to work around it.

i see,lol

**Comment by Justaus3r** [comment]
Here are search results with default provider:
image

Where's here are results with animerush:
image
its showing me the season i wanna watch at no 1 index.

**Comment by RaitaroH** [comment]
Provider aside, I tested on my own. It seems that `mpv.exe` is a GUI application which does **not** output anything in the terminal. Therefore `adl` will not be able to count any episodes being watched. If not otherwise specified, adl will default to `mpv.exe` on my machine.
On the other hand `mpv.com` is what we need to run in the terminal as player to actually get any output, **but** on my end I get some ydl problems.
Can you confirm you can go through the whole process (start adl, watch an episode, update ep count) without any problems?

<details>
<summary>using mpv.exe by default - no mpv output, thus no counting</summary>

```
Tsuki ga Michibiku Isekai Douchuu > 9 / 12
Enter lowercase or uppercase to issue command:
      N - Next episode (default, press <ENTER>)
      L - from current to Last known
      A - All available, from episode 1
      I - custom Interval (range) of episodes
    0-9 - Plus n episodes relative to last seen (type number)
      R - Rewatch/redownload current episode in list
      C - Custom episode
      U - Update entry chosen instead of streaming
      S - Skip. Choose another show.
Your choice? [N/l/a/i/0-9/r/c/u/s]:

Now watching Tsuki ga Michibiku Isekai Douchuu, episode #10
You can press Ctrl+C to cancel watching. That episode won't be counted.
anime: anime-downloader 5.0.14
+--------+--------------------------------------------------------------------------------+--------+
[...]
|      2 | Mondaiji-tachi ga Isekai kara Kuru Sou Desu yo? OVA                            |        |
|      1 | Tsuki ga Michibiku Isekai Douchuu                                              |        |
|--------+--------------------------------------------------------------------------------+--------|
|   SlNo | Title                                                                          | Meta   |
+--------+--------------------------------------------------------------------------------+--------+
Enter the anime no: [1]: 1
anime: Selected Tsuki ga Michibiku Isekai Douchuu
anime: Extracting episode info from page
anime: Found anime: Tsuki ga Michibiku Isekai Douchuu
anime: Speed is being limited to 0

mpv didn't start, the anime wasn't found, or the episode wasn't found...
Try another provider? [N / provider name]: n
Skipping different provider retry...
Proceeding...

Want to watch another anime? [Y/n]: n
Skipping watching another...

Thank you for using adl ✨!
Exiting.

```
</details>
<details>
<summary>using mpv.com - video doesn't start, but that might be fixable - trackma update works</summary>

```
Tsuki ga Michibiku Isekai Douchuu > 9 / 12
Enter lowercase or uppercase to issue command:
      N - Next episode (default, press <ENTER>)
      L - from current to Last known
      A - All available, from episode 1
      I - custom Interval (range) of episodes
    0-9 - Plus n episodes relative to last seen (type number)
      R - Rewatch/redownload current episode in list
      C - Custom episode
      U - Update entry chosen instead of streaming
      S - Skip. Choose another show.
Your choice? [N/l/a/i/0-9/r/c/u/s]:

Now watching Tsuki ga Michibiku Isekai Douchuu, episode #10
You can press Ctrl+C to cancel watching. That episode won't be counted.
anime: anime-downloader 5.0.14
+--------+--------------------------------------------------------------------------------+--------+
[...]
|      2 | Mondaiji-tachi ga Isekai kara Kuru Sou Desu yo? OVA                            |        |
|      1 | Tsuki ga Michibiku Isekai Douchuu                                              |        |
|--------+--------------------------------------------------------------------------------+--------|
|   SlNo | Title                                                                          | Meta   |
+--------+--------------------------------------------------------------------------------+--------+
Enter the anime no: [1]:
anime: Selected Tsuki ga Michibiku Isekai Douchuu
anime: Extracting episode info from page
anime: Found anime: Tsuki ga Michibiku Isekai Douchuu
anime: Speed is being limited to 0
[ffmpeg] https: HTTP error 403 Forbidden
Failed to open https://www3.mp4upload.com:282/d/r2x2llouz3b4quuomkxr6ikxi5upqwea3lawvaw75l3oqhhhhgvmnxos/video.mp4.
[ytdl_hook] ERROR: Unable to download webpage: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)> (caused by URLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)')))
[ytdl_hook] youtube-dl failed: unexpected error occurred

Exiting... (Errors when loading file)

Want to further watch the anime? Enter number for ep. count, [N/0-9]: n
Proceeding...
Increase nr in anime list by 1? Yes, no, or custom number [Y/n/0-9]: y
←[0;32mEngine: Updating show Tsuki ga Michibiku Isekai Douchuu to episode 10...←[0m
←[0;33mData: Queued update for Tsuki ga Michibiku Isekai Douchuu←[0m

Want to watch another anime? [Y/n]: n
Skipping watching another...

Thank you for using adl ✨!
Exiting.
```
</details>

**Comment by Justaus3r** [comment]
>> Provider aside, I tested on my own. It seems that mpv.exe is a GUI application which does not output anything in the terminal. Therefore adl will not be able to count any episodes being watched. If not otherwise specified, adl will default to mpv.exe on my machine.
On the other hand mpv.com is what we need to run in the terminal as player to actually get any output, but on my end I get some ydl problems.

Yeah,it looks like they are using ```.exe``` format for GUI and ```.com``` format for command-line.**BUT** it should not be a problem since when you install  mpv from chocolatey it adds both ```mpv.exe``` and ```mpv.com``` on PATH and typing ```mpv``` on cmd will refer to ```mpv.com```.that is unless you are using git-bash instead of cmd/powershell.in this case you will have to explicitly use ```mpv.com``` to access mpv and typing mpv will result in command not found error.so if you are using git-bash for executing adl then don't,instead add bash.exe to PATH and execute script as ```bash .\adl``` from cmd/powershell.

>> Can you confirm you can go through the whole process (start adl, watch an episode, update ep count) without any problems?

Yes its working perfectly on both windows 7 system and windows 10 system.
<details>
<summary>Take a look at screenshots</summary>
<br>

Screenshot_1

Screenshot_2

Screenshot_3

Screenshot_4

Screenshot_5

</details>

>> using mpv.com - video doesn't start, but that might be fixable - trackma update works

Yes about that,it seems to be a validation problem in ssl certificate.try changing the provider to something that doesn't require ffmpeg. 

Also please take a note that the default provider in my fork is animerush which requires nodejs to be installed,which btw you can do with ```choco install nodejs```.

Maybe for further conversation we can use discord,if you are fine with it that is:).my discord: Neko\~Neko\~ni#3050

**Comment by RaitaroH** [comment]
As mentioned on my end adl defaults to mpv.exe, not mpv.com. Anyhow specifically using `mpv.com` does work with a provider like `animeout` for example which is good. We could just change the var to `mpv.com` to make sure, maybe under `echo "Platform:Windows"` or something. Most likely I will just specify in the README.

Furthermore, I noticed that your batch script only tests for mpv. I changed it so it can take a variable to test instead. Also, a bit of a wait for the script to run is needed.

**Comment by RaitaroH** [comment]
This PR will definitely be squashed 🤭

File: adl [hunk 1]
```diff
@@ -51,11 +52,26 @@ reset_vars() { #{{{
...
+  if ! [ -x "$(command -v $windows_specific_cmd)" ]; then
+    echo "Platform:Linux"
+    if ! [ -x "$(command -v $player)" ]; then
+      echo "Error: '$player' is not installed." >&2
+      print_options
+      exit 1
+    fi
+  else 
+    echo "Platform:Windows"
+    [ "$player" == "mpv" ] && player="mpv.com"
```
Comment (RaitaroH) [inline adl]: @Justaus3r this line should take care of it.

File: adl [hunk 2]
```diff
@@ -230,8 +246,7 @@ animedl() { #{{{
     choice_arg()   { [[ -z "$yes" ]] && echo "" || echo "-c 1" ;}
     player_arg="$1"
     down_arg()     { [[ "$download" == "0" ]] && echo "--play $player_arg"   || echo "" ;}
-    provider_arg() { [[ "$provider" != ""  ]] && echo "--provider $provider" || echo "" ;}
-
+    provider_arg() { [[ "$provider" != ""  ]] && echo "--provider $provider" || echo "--provider animerush" ;} # setting default provider to animerush because it gives far better results
```
Comment (RaitaroH) [inline adl]: I would most likely revert this, as it has been covered by readme. Another option is to just set this at the [top](https://github.com/RaitaroH/adl/blob/master/adl#L30) of the script.

File: adl (hunk 1 above)
Comment (Justaus3r) [reply adl:65]: Yeah you can do that ,we can let users decide which provider they wanna use

File: adl (hunk 2 above)
Comment (Justaus3r) [reply adl]: Okie ,we can let users decide which provider they wanna use

**Comment by Justaus3r** [comment]
>> As mentioned on my end adl defaults to mpv.exe, not mpv.com. 

That is because(probably but maybe i might be wrong) in your %PATHEXT% ```.exe``` extension has more precedence than ```.com```,see here:
```
$ echo %PATHEXT%
.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC
```
starting from order is the precedence in which files of same name are executed if the extension isn't explicitly provided,what i think is your %PATHEXT% might be something like:
```
.EXE;.COM;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC
```
if this is the case then edit the %PATHEXT% variable and give ```.COM``` more precedence than ```.EXE```.

**BUT for most of the user's this shoudn't be a problem as ```.COM``` has more precedence than ```.EXE``` by default.we can add a note to readme telling users to check the %PATHEXT% variable before/after downloading mpv.**

>> Anyhow specifically using mpv.com does work with a provider like animeout for example which is good.

I dunno if am stupid but i really don't see how a provider not working could be a _**good**_ thing in anyway 😅 .

>> We could just change the var to mpv.com to make sure, maybe under echo "Platform:Windows" or something. Most likely I will just specify in the README.

Yes about that,i would have agreed with you but seeing as specifying ```mpv.com``` leads to failure of some providers,i am against this change as this leads to failure of many providers.i think most of users won't face this issue and even if one does face this issue we can add a note in readme or create a issue and pin it,telling them to check if ```mpv.com``` is present in ```C:\ProgramData\chocolatey\lib\mpv.install\tools``` and they have ```.COM``` more precedent-ed than ```.EXE```.like why restrict users when they can use any provider 😃.i actually cloned my fork after you made these changes and yeah many providers like animeout and animerush were not working and i had to change the player from ```mpv.com``` to ```mpv```.i haven't reverted the ```mpv.com``` change as i would like to hear your thoughts on the matter.

Mpv actually has an [issue](https://github.com/mpv-player/mpv/issues/8544) regarding this matter.reading that might help.

>> Furthermore, I noticed that your batch script only tests for mpv. I changed it so it can take a variable to test instead. Also, a bit of a wait for the script to run is needed.

Ah yes thanks i didn't notice that there was option to change the player.while i would have preferred to avoid calling another executable(where.exe),this is fine too.

File: adl (hunk 1 above)
Comment (Justaus3r) [reply adl]: i am afraid this leads to failure of many providers  😔

**Comment by RaitaroH** [comment]
Ok. It is not worth to fix one problem to cause 10 more so we will stick to `mpv`. We can formulate something in the README regarding `mpv.com`, PATH and what not. But really important is if `adl` actually is able to count the episodes being watched properly. So if you are watching say 2 ep, then when asked input 1 to watch a 3rd episode,  will adl count 3 under your system? Or not. If the counting doesn't work with `mpv` then I suppose we will have to either accept the loss of provider options.

> I dunno if am stupid but i really don't see how a provider not working could be a good thing in anyway sweat_smile .

Yeah, I meant that at least I can verify that adl works on my machine.

> Ah yes thanks i didn't notice that there was option to change the player.while i would have preferred to avoid calling another executable(where.exe),this is fine too.

If you can find a better way to do it then please do. I don't know much about batch scripting. We could even use a ps1 script. I have seen that you can just do basic IFs instead of the calls you used. Is there a reason behind that? 
Also also, it might be a bit suspicious to people that a CMD window is briefly opened when testing the player. Is that avoidable?

**Comment by Justaus3r** [comment]
>> Ok. It is not worth to fix one problem to cause 10 more so we will stick to mpv. We can formulate something in the README regarding mpv.com, PATH and what not. But really important is if adl actually is able to count the episodes being watched properly. So if you are watching say 2 ep, then when asked input 1 to watch a 3rd episode, will adl count 3 under your system? Or not. If the counting doesn't work with mpv then I suppose we will have to either accept the loss of provider options.

Episode counting works fine.

>> Yeah, I meant that at least I can verify that adl works on my machine.

Oh i see

>> If you can find a better way to do it then please do. I don't know much about batch scripting. 

Well i kinda i have experience with batch-scripting,but to be honest there are alot of limitations in batch-files.but i will see what i can do.

>> We could even use a ps1 script

i would have if i could,but i don't have experience scripting with powershell.

>> I have seen that you can just do basic IFs instead of the calls you used. Is there a reason behind that?

Yeah we can do that but it's my habit of creating functions.the calls are not necessary.and since you mentioned it i will change that. 

>> Also also, it might be a bit suspicious to people that a CMD window is briefly opened when testing the player. Is that avoidable?

Well yeah it is possible we can just run the batch-file hidden but know that an additional .vbs file will be added to codebase

**Comment by RaitaroH** [comment]
Ok then. You can change the batch script a bit and I will see about the readme.

> Well yeah it is possible we can just run the batch-file hidden but know that an additional .vbs file will be added to codebase

Let's not then. Simpler is better.

**Comment by Justaus3r** [comment]
>> Let's not then. Simpler is better.

Well its not that complex. a simple .vbs script is needed,there is no need to install software for that as .vbs files can execute natively on windows.But thats up to you,what do you say?

**Comment by Justaus3r** [comment]
btw the batch file is now as simple as this:
```batchfile
@echo off
setlocal enabledelayedexpansion
set player_name=%~1
%player_name% >nul 2>&1
if %errorlevel% neq 0 (echo 1 >player_bool_file && exit) else (call echo 0 >player_bool_file && exit)
```
btw can you run my fork now?

**Comment by RaitaroH** [comment]
> Well its not that complex. a simple .vbs script is needed,there is no need to install software for that as .vbs files can execute natively on windows.But thats up to you,what do you say?

I mean, I can test it and then decide.

> btw can you run my fork now?

I will.

**Comment by Justaus3r** [comment]
>> I mean, I can test it and then decide.

OK i will push it to my fork you can test it and if you are satisfied then you can merge it.

**Comment by RaitaroH** [comment]
I tested the batch, but isn't it backwards? Here is the edit:

```bat
@echo off
setlocal enabledelayedexpansion
set player_name=%~1
%player_name% >nul 2>&1
if %errorlevel% neq 0 (echo 0 >player_bool_file && exit) else (echo 1 >player_bool_file && exit)
```
Will result now in:
```
$ .\player_check.bat 'mpv' ; cat .\player_bool_file
1
$ .\player_check.bat 'vlc' ; cat .\player_bool_file
0
```

Also, yeah about that PATHEXT...
```
$ echo %PATHEXT%
.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.PY;.PYW
```

**Comment by Justaus3r** [comment]
>> I tested the batch, but isn't it backwards? Here is the edit:

Sorry it was a typo

>> Also, yeah about that PATHEXT...
$ echo %PATHEXT%
.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.PY;.PYW

Well your PATHEXT variable seems to be fine.
if the problem persists,try reinstalling mpv from choco again..

Edit:If you have a discord account,we could talk on that as this Pr has kinda become a mess
Edit2:Since `anime dl` only supports vlc and mpv for streaming can i hardcode them?

**Comment by RaitaroH** [comment]
@all-contributors please add @Justaus3r  for code, tutorial

**Pull request merged**